src/librados/IoCtxImpl.cc

```diff
@@ -1235,7 +1235,7 @@ int librados::IoCtxImpl::remove(const object_t& oid)
   ::ObjectOperation op;
   prepare_assert_ops(&op);
   op.remove();
-  return operate(oid, &op, nullptr, librados::OPERATION_FULL_FORCE);
+  return operate(oid, &op, nullptr, CEPH_OSD_FLAG_FULL_FORCE);
```

Comment:
Let's take a look on the flow:

```cpp
int librados::IoCtxImpl::operate(const object_t& oid, ::ObjectOperation *o,
                                 ceph::real_time *pmtime, int flags)
{
  // ...
  Objecter::Op *objecter_op = objecter->prepare_mutate_op(
    oid, oloc,
    *o, snapc, ut,
    flags | extra_op_flags,
    oncommit, &ver);
```

```cpp
  // mid-level helpers
  Op *prepare_mutate_op(
    const object_t& oid, const object_locator_t& oloc,
    ObjectOperation& op, const SnapContext& snapc,
    ceph::real_time mtime, int flags,
    Context *oncommit, version_t *objver = NULL,
    osd_reqid_t reqid = osd_reqid_t(),
    ZTracer::Trace *parent_trace = nullptr) {
    Op *o = new Op(oid, oloc, std::move(op.ops), flags | global_op_flags |
                   CEPH_OSD_FLAG_WRITE, oncommit, objver,
                   nullptr, parent_trace);
```

The union with `CEPH_OSD_FLAG_WRITE` suggests the conceptual type.

All constructors of `Op` do:

```cpp
  struct Op : public RefCountedObject {
    // ...
    op_target_t target;
    // ...
    Op(const object_t& o, const object_locator_t& ol,  osdc_opvec&& _ops,
       int f, /* ... */) :
      target(o, ol, f),
```

```cpp
  struct op_target_t {
    int flags = 0;
    // ...
    op_target_t(const object_t& oid, const object_locator_t& oloc, int flags)
      : flags(flags),
    // ...
    bool respects_full() const {
      return
        (flags & (CEPH_OSD_FLAG_WRITE | CEPH_OSD_FLAG_RWORDERED)) &&
        !(flags & (CEPH_OSD_FLAG_FULL_TRY | CEPH_OSD_FLAG_FULL_FORCE));
    }
```

`respects_full()` clearly assumes `CEPH_OSD_FLAG_FULL_FORCE`.

Reply:
The `respects_full()` governs the objecter's behavior on traffic pausing:

```cpp
bool Objecter::target_should_be_paused(op_target_t *t)
{
  const pg_pool_t *pi = osdmap->get_pg_pool(t->base_oloc.pool);
  bool pauserd = osdmap->test_flag(CEPH_OSDMAP_PAUSERD);
  bool pausewr = osdmap->test_flag(CEPH_OSDMAP_PAUSEWR) ||
    (t->respects_full() && (_osdmap_full_flag() || _osdmap_pool_full(*pi)));

  return (t->flags & CEPH_OSD_FLAG_READ && pauserd) ||
    (t->flags & CEPH_OSD_FLAG_WRITE && pausewr) ||
    (osdmap->get_epoch() < epoch_barrier);
}
```

However, the main responsibility of `op_target_t::flags` is being passed to an OSD via `MOSDOp`:

```cpp
Objecter::MOSDOp *Objecter::_prepare_osd_op(Op *op)
{
  // rwlock is locked

  int flags = op->target.flags;
  flags |= CEPH_OSD_FLAG_KNOWN_REDIR;
  flags |= CEPH_OSD_FLAG_SUPPORTSPOOLEIO;

  // Nothing checks this any longer, but needed for compatibility with
  // pre-luminous osds
  flags |= CEPH_OSD_FLAG_ONDISK;

  if (!honor_pool_full)
    flags |= CEPH_OSD_FLAG_FULL_FORCE;

  // ...
  auto m = new MOSDOp(client_inc, op->tid,
                      hobj, op->target.actual_pgid,
                      osdmap->get_epoch(),
                      flags, op->features);
```